kv/util/util.go

```diff
@@ -32,55 +32,144 @@ import (
 
 var errNilStore = errors.New("kv store is nil")
 
+// ValidateFn validates an update from KV.
+type ValidateFn func(interface{}) error
+
 type getValueFn func(kv.Value) (interface{}, error)
+
 type updateFn func(interface{})
 
 // WatchAndUpdateBool sets up a watch for a bool property.
-func WatchAndUpdateBool(store kv.Store, key string, property *bool, lock sync.Locker,
-	defaultValue bool, logger xlog.Logger) (kv.ValueWatch, error) {
+func WatchAndUpdateBool(
+	store kv.Store,
+	key string, property *bool,
+	lock sync.Locker,
+	defaultValue bool,
+	logger xlog.Logger,
+) (kv.ValueWatch, error) {
 	updateFn := lockedUpdate(func(i interface{}) { *property = i.(bool) }, lock)
 
-	return watchAndUpdate(store, key, getBool, updateFn, defaultValue, logger)
+	return watchAndUpdate(store, key, getBool, updateFn, nil, defaultValue, logger)
+}
+
+// WatchAndUpdateFloat64 sets up a watch for a float64 property.
+func WatchAndUpdateFloat64(
+	store kv.Store,
+	key string,
+	property *float64,
+	lock sync.Locker,
+	defaultValue float64,
+	logger xlog.Logger,
+) (kv.ValueWatch, error) {
+	return WatchAndUpdateWithValidationFloat64(store, key, property, lock, defaultValue, nil, logger)
+}
+
+// WatchAndUpdateInt64 sets up a watch for an int64 property.
+func WatchAndUpdateInt64(
+	store kv.Store,
+	key string,
+	property *int64,
+	lock sync.Locker,
+	defaultValue int64,
+	logger xlog.Logger,
+) (kv.ValueWatch, error) {
+	return WatchAndUpdateWithValidationInt64(store, key, property, lock, defaultValue, nil, logger)
+}
+
+// WatchAndUpdateString sets up a watch for a string property.
+func WatchAndUpdateString(
+	store kv.Store,
+	key string,
+	property *string,
+	lock sync.Locker,
+	defaultValue string,
+	logger xlog.Logger,
+) (kv.ValueWatch, error) {
+	return WatchAndUpdateWithValidationString(store, key, property, lock, defaultValue, nil, logger)
+}
+
+// WatchAndUpdateTime sets up a watch for a time property.
+func WatchAndUpdateTime(
+	store kv.Store,
+	key string,
+	property *time.Time,
+	lock sync.Locker,
+	defaultValue time.Time,
+	logger xlog.Logger,
+) (kv.ValueWatch, error) {
+	return WatchAndUpdateWithValidationTime(store, key, property, lock, defaultValue, nil, logger)
 }
 
-// WatchAndUpdateFloat64 sets up a watch for an float64 property.
-func WatchAndUpdateFloat64(store kv.Store, key string, property *float64, lock sync.Locker,
-	defaultValue float64, logger xlog.Logger) (kv.ValueWatch, error) {
+// WatchAndUpdateWithValidationFloat64 sets up a watch with validation for a float64
+// property. Any invalid updates are not applied.
+func WatchAndUpdateWithValidationFloat64(
```

Comment:
Can you do a search across OSS and internal repos to see how many places you'd need to change if users are always required to pass in validation function (`nil` if no validation function)? Having both versions maintains API backward compatibility at the cost of wider and more verbose public interface so I'd prefer to just merge them if it doesn't take too much effort.

Reply:
Yea, meant to follow up on this after standup, I think that's the better approach, will take a look across OSS and statsdex.